src/riak_core_bucket.erl

```diff
@@ -192,14 +197,61 @@ bucket_nval_map(Ring) ->
 default_object_nval() ->
     riak_core_bucket:n_val(riak_core_bucket_props:defaults()).
 
+%% @private
+%% @doc For testing/debugging
+-spec all_n() -> [pos_integer(),...].
+all_n() ->
+    {ok, Ring} = riak_core_ring_manager:get_my_ring(),
+    all_n(Ring).
+
+-spec all_n(riak_core_ring()) -> [pos_integer(),...].
 all_n(Ring) ->
-    BucketNs = bucket_nval_map(Ring),
-    DefaultN = default_object_nval(),
-    AllN = lists:foldl(fun({_, N}, Acc) ->
-                               ordsets:add_element(N, Acc)
-                       end, [DefaultN], BucketNs),
+    BucketNVals = bucket_nvals(Ring),
+    BucketTypeNVals = bucket_type_nvals(),
+    ordsets:from_list(
+        ordsets:union(BucketNVals, BucketTypeNVals)
+    ).
+
+%% @private
+-spec bucket_nvals(riak_core_ring()) -> nval_set().
+bucket_nvals(Ring) ->
+    Buckets = riak_core_ring:get_buckets(Ring),
+    BucketProps = [get_bucket(Bucket, Ring)
+        || Bucket <- Buckets],
+    Default = app_helper:get_env(riak_core, default_bucket_props),
+    DefaultN = proplists:get_value(n_val, Default),
+    AllN = lists:foldl(
+        fun bucket_prop_nval/2,
+        ordsets:from_list([DefaultN]), BucketProps),
     AllN.
 
+%% @private
+-spec bucket_type_nvals() -> nval_set().
+bucket_type_nvals() ->
+    riak_core_bucket_type:fold(fun bucket_type_prop_nval/2, ordsets:new()).
+
+%% @private
+-spec bucket_type_prop_nval({bucket_type(), bucket_props()}, nval_set()) ->
+    nval_set().
+bucket_type_prop_nval({_BType, BProps}, Accum) ->
+    case proplists:get_value(active, BProps) of
+        true ->
+            bucket_prop_nval(BProps, Accum);
+        _ ->
+            Accum
+    end.
+
+%% @private
+-spec bucket_prop_nval(bucket_props(), nval_set()) ->
+    nval_set().
+bucket_prop_nval(BProps, Accum) ->
+    case proplists:get_value(n_val, BProps) of
```

Comment:
Not sure if it's worth bothering to change, since this may not be a performance-critical path, but the `lists:keyfind` function is substantially faster than `proplists:get_value`. The functions in the `proplists` module have additional features that make them worthwhile sometimes, but for this case `lists:keyfind` has pretty much the same semantics anyway.